File: ios/Infrastructure/MLKit/MLKitTextRecognitionService.swift
import Foundation

#if MLKIT_TEXT_RECOGNITION
  import MLKitTextRecognition
#endif

#if MLKIT_TEXT_RECOGNITION_CHINESE
  import MLKitTextRecognitionChinese
#endif

#if MLKIT_TEXT_RECOGNITION_DEVANAGARI
  import MLKitTextRecognitionDevanagari
#endif

#if MLKIT_TEXT_RECOGNITION_JAPANESE
  import MLKitTextRecognitionJapanese
#endif

#if MLKIT_TEXT_RECOGNITION_KOREAN
  import MLKitTextRecognitionKorean
#endif

#if MLKIT_TEXT_RECOGNITION
  class MLKitTextRecognitionService: IRecognitionService {
    typealias ResultType = TextRecognitionResult

    private let textRecognizer: TextRecognizer

    init(textRecognizer: TextRecognizer) {
      self.textRecognizer = textRecognizer
    }

    func recognize(image: ProcessedImage) throws -> TextRecognitionResult {
      let text = try textRecognizer.results(in: image.image)
      return MLKitTextAdapter.toDomain(text)
    }
  }
#endif

File: android/src/main/java/com/visioncameramlkit/infrastructure/mlkit/MLKitTextRecognitionService.kt
package com.visioncameramlkit.infrastructure.mlkit

import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognizer
import com.visioncameramlkit.domain.models.ProcessedImage
import com.visioncameramlkit.domain.models.TextRecognitionResult
import com.visioncameramlkit.domain.services.IRecognitionService
import com.visioncameramlkit.infrastructure.mlkit.adapters.MLKitTextAdapter

class MLKitTextRecognitionService(
  private val textRecognizer: TextRecognizer,
) : IRecognitionService<TextRecognitionResult> {
  // ML Kit's on-device detectors are not documented as safe for concurrent use.
  // This instance is shared between the synchronous frame-processor path and the
  // async static-image path, so calls (and disposal) must be serialized.
  private val lock = Any()

  override fun recognize(image: ProcessedImage): TextRecognitionResult =
    synchronized(lock) {
      val task = textRecognizer.process(image.image)
      val mlkitText: Text? = Tasks.await(task)
      MLKitTextAdapter.toDomain(mlkitText)
    }

  override fun close() {
    synchronized(lock) {
      textRecognizer.close()
    }
  }
}
